[Feat/#7] ApiGatewayHandler 구현

JungYoonShin opened the pull request:
<!--
name: Makers Push Notification Feature PR template
about: 구현한 기능과 변경 사항에 대해 구체적으로 작성해주세요
title: '[FEAT/{#이슈번호}] 기능 내용'
labels: ''
assignees: ''
-->

<!--
- 리뷰어 추가했나요?
- 허가자 추가했나요?
- 라벨 추가했나요?
-->

## Related Issue 🚀
- closed #17

## Work Description ✏️
- APiGatewayHandler 구현했습니다!!
- DynamoDB에 토큰 등록, 삭제 잘 되는 것 확인했고, 알림도 잘 오는 것 확인했습니다!!
- 테스트 구현 방법은 [노션](https://www.notion.so/sopt-makers/ApiGateWayHandler-2d076042aac280628e5ee5838135d8fd?source=copy_link)에 정리해두었어요 !


### ApiGatewayHandler 구현 요약

>  REGISTER, CANCEL, SEND, SEND_ALL 총 4가지 기능

🔆 액션별 처리 내용

1) REGISTER: DeviceTokenService.registerToken() 호출 + 히스토리 로깅
2) CANCEL: InvalidEndpointCleaner.clean()으로 통합 정리 (DynamoDB 삭제 + SNS 리소스 정리)
3) SEND: SendPushFacade.sendPush() 호출
4) SEND_ALL: SendPushFacade.sendPushAll() 호출

## PR Point 📸
<!-- 피드백을 받고 싶은 부분을 적어주세요. -->

JungYoonShin committed:
[FEAT] DeviceTokenException 추가

- 관련 ErrorMessage 추가

JungYoonShin committed:
[REFACTOR] 기본 생성자 롬복으로 대체

JungYoonShin committed:
[FIX] DeviceTokenService 구현(토큰 등록,  유저 저장)

JungYoonShin committed:
[FEAT] pk 기반 디바이스 토큰 조회 추가

JungYoonShin committed:
[FIX] 추가 구현에 따른 AppFactory 수정

JungYoonShin committed:
[FEAT] SnsFactory 구현

- sns 구독, endpoint ARN 등록 등

JungYoonShin committed:
[FEAT] 토큰 등록 & 삭제 관련 DTO 구현

JungYoonShin committed:
[FEAT] 알림 전송 관련 DTO 구현

JungYoonShin committed:
[FEAT] ApiGatewayHandler 구현

JungYoonShin committed:
[FIX] template.yml 수정

ApiGatewayHandler 클래스명과 함수명 수정

JungYoonShin committed:
[MERGE] 충돌 해결

JungYoonShin committed:
[MERGE] 충돌 머지

JungYoonShin committed:
[REFACTOR] 불필요한 파일 삭제

JungYoonShin committed:
[REFACTOR] 코드 스타일 적용

seoyeonjin commented on src/main/java/com/sopt/push/common/PushFailException.java:
```diff
@@ -0,0 +1,8 @@
+package com.sopt.push.common;
+
+public class PushFailException extends RuntimeException {
```
이거 지웠었는데 머지하면서 다시 들어간 거 같아요..!
예외처리는 ExternalException으로 변경했습니다

JungYoonShin replied:
요것두 다시 확인할게요 !!

seoyeonjin commented on src/main/java/com/sopt/push/config/AppFactory.java:
```diff
@@ -46,9 +46,17 @@ private AppFactory() {
 
+    HistoryService historyService = new HistoryService(historyRepository);
+    NotificationService notificationService = new NotificationService(snsClient, envConfig);
+    SnsFactory snsFactory = new SnsFactory(snsClient, envConfig);
```
snsFactory를 따로 만든 이유가 있나요? 
notificationService에서 sns 관련 로직 담당하고 있어서 snsFactory는 또 만들 필요는 없을 거 같습니다!

JungYoonShin replied:
요거 의논한대루 notificationService 로 옮겨놓을게요 !!

seoyeonjin commented on src/main/java/com/sopt/push/config/AppFactory.java:
```diff
@@ -46,9 +46,17 @@ private AppFactory() {
     HistoryRepository historyRepository = new HistoryRepository(dynamoClient, tableName);
     DeviceTokenRepository tokenRepository = new DeviceTokenRepository(dynamoClient, tableName);
 
+    HistoryService historyService = new HistoryService(historyRepository);
```
historyService는 따로 사용 안 하는 거 같은데 추가한 이유가 있을까요??

JungYoonShin replied:
엇 그렇네요 제거할게요!! 감사해요!

seoyeonjin commented on src/main/java/com/sopt/push/config/AppFactory.java:
```diff
@@ -46,9 +46,17 @@ private AppFactory() {
+    DeviceTokenService deviceTokenService =
+        new DeviceTokenService(tokenRepository, userRepository, snsFactory);
+    UserService userService = new UserService(userRepository);
+    InvalidEndpointCleaner invalidEndpointCleaner =
```
아래 `this.invalidEndpointCleaner = new InvalidEndpointCleaner()` 로 만들고 있어서 또 만들지 않아도 괜찮을 거 같습니다 !

JungYoonShin replied:
예리한 캐치 감사합니다 !!

seoyeonjin commented on src/main/java/com/sopt/push/config/AppFactory.java:
```diff
@@ -46,9 +46,17 @@ private AppFactory() {
     this.userService = new UserService(userRepository);
     this.historyService = new HistoryService(historyRepository);
-    this.deviceTokenService = new DeviceTokenService(tokenRepository);
+    this.deviceTokenService = new DeviceTokenService(tokenRepository, userRepository, snsFactory);
```
user 관련 로직은 userService 로 옮겨주실 수 있나요?
deviceTokenService 와 UserService 를 분리해서 만들어놓았습니당
```
deviceTokenService - deviceRepository
userService - userRepository 
```

JungYoonShin replied:
넵 한번 리팩해볼게요 ~ !

seoyeonjin commented on src/main/java/com/sopt/push/dto/LogCreateRequestDto.java:
```diff
@@ -0,0 +1,39 @@
+import com.sopt.push.enums.Services;
+import java.util.Set;
+
+public record LogCreateRequestDto(
```
기존에 CreateHistoryDto 가 있어서 그걸로 사용하면 좋을 거 같아요

seoyeonjin commented on src/main/java/com/sopt/push/dto/SendPushDto.java:
```diff
@@ -0,0 +1,12 @@
+import com.sopt.push.enums.Category;
+import java.util.Set;
+
+public record SendPushDto(
```
기존에 만들어두었던 `RequestSendPushMessageDto`가 있어서 그걸로 사용해도 될 거 같아요

JungYoonShin replied:
넵 !!

seoyeonjin commented on src/main/java/com/sopt/push/dto/SendAllPushDto.java:
```diff
@@ -0,0 +1,6 @@
+
+import com.sopt.push.enums.Category;
+
+public record SendAllPushDto(
```
기존에 만들어두었던 `RequestSendAllPushMessageDto` 가 있어서 그걸로 사용해도 될 것 같아요

seoyeonjin commented on src/main/java/com/sopt/push/lambda/ApiGatewayHandler.java:
```diff
@@ -0,0 +1,335 @@
+      return convertToApiGatewayResponse(responseMap);
+
+    } catch (BusinessException ex) {
+      context.getLogger().log("ApiGateway error: " + ex.getMessage());
```
저희 로깅 방식 `@Slf4j` 로 통일하기로 해서 그걸로 맞춰줄 수 있을까영?

JungYoonShin replied:
넵넵 !!

seoyeonjin commented on src/main/java/com/sopt/push/lambda/ApiGatewayHandler.java:
```diff
@@ -0,0 +1,335 @@
+  private ApiGatewayRequestDto extractRequest(APIGatewayProxyRequestEvent event) {
+    Map<String, String> headers = event.getHeaders();
+
+    if (headers == null || headers.get("action") == null) {
```
하드코딩된 부분 상수화하면 어떨까용 ??

JungYoonShin replied:
넵 좋아요 감사합니다 !!

sung-silver replied:
여기도 isInvalidHeader로 변수 추출하면 좋을 것 같아요!

JungYoonShin replied:
넵 !!

seoyeonjin commented on src/main/java/com/sopt/push/lambda/ApiGatewayHandler.java:
```diff
@@ -0,0 +1,335 @@
+    validate(finalDto);
+
+    try {
+      deviceTokenService.registerToken(deviceToken, platform, userId);
```
deviceTokenService에서 토큰 등록, 사용자 등록, 엔드포인트 등록까지 하고 있는 거 맞나요?
기존에 역할이 나눠져있었어서
```
deviceTokenService.registerToken
userService.registerUser
notificationService.subscribe ... 
```
이런 식으로 나눠서 호출하면 좋을 것 같아요

`InvalidEndpointCleaner` 처럼 이 로직들 묶어서 한번에 호출하도록 분리해도 좋을 것 같습니당

JungYoonShin replied:
넵 좋아요 !! 제안 너무 감사합니다 ~!

seoyeonjin commented on src/main/java/com/sopt/push/lambda/ApiGatewayHandler.java:
```diff
@@ -0,0 +1,335 @@
+    Platform platform = request.header().platform();
+    String deviceToken = body.deviceToken();
+    Set<String> userIds = body.userIds();
+    Set<String> logUserIds = Set.of("NULL");
```
userIds를 사용하지 않고 logUserIds를 따로 만드는 이유가 있나요 ??

JungYoonShin replied:
앗 요거 코드가 조금 잘못 짠 것 같아요..! userIds로 사용할게요 감사해요 !!

seoyeonjin commented on src/main/java/com/sopt/push/lambda/ApiGatewayHandler.java:
```diff
@@ -0,0 +1,335 @@
+        throw new DeviceTokenException(ErrorMessage.ARN_UNDEFINED);
+      }
+
+      String actualUserId = extractUserIdFromSk(tokenEntity.getSk());
```
prefix 제거 안 해도 됩니당
db에도 prefix 추가된 상태로 저장되어있는 거라서 괜찮습니다!

JungYoonShin replied:
앗 요기 이거 제거한 이유가 `invalidEndpointCleaner.clean()`에서 아래 코드를 호출하는데 이 메서드들이 prefix를 붙여서 제거하길래 요기서 제거한거였습니당 !!

`    tokenService.deleteToken(token.userId(), token.deviceToken());`

seoyeonjin replied:
제가 기존에 제거하는 코드를 넣었었네요 ㅠ 
이 부분은 이대로 구현해주시면 될 거 같습니다!
지금 생각엔 prefix 제거할 필요가 딱히 없는 거 같아서 후에 리팩하겠습니다

JungYoonShin replied:
넵 !! 좋습니당

seoyeonjin commented on src/main/java/com/sopt/push/lambda/ApiGatewayHandler.java:
```diff
@@ -0,0 +1,335 @@
+    return pk;
+  }
+
+  private void createRegisterLog(
```
createRegisterLog랑 createCancelLog 가 인자만 다른 거 같은데 하나로 합치는 건 어떤가요?

JungYoonShin replied:
넵 반영하겠습니당 !!

seoyeonjin commented on src/main/java/com/sopt/push/repository/DeviceTokenRepository.java:
```diff
@@ -32,6 +33,12 @@ public Optional<DeviceTokenEntity> findByPkAndSk(String pk, String sk) {
     return Optional.ofNullable(deviceTokenTable.getItem(key));
   }
 
+  public List<DeviceTokenEntity> queryByPk(String pk) {
```
현재는 네이밍을 pk -> DeviceToken으로 하고 있어서 함수명을 `findAllByDeviceToken` 같이 바꾸면 의미가 더 잘 전달될 거 같아요

JungYoonShin replied:
넵 좋아욧!!!

seoyeonjin commented on src/main/java/com/sopt/push/service/NotificationService.java:
```diff
@@ -66,9 +66,9 @@ public String platformPush(
 
     } catch (SnsException ex) {
-      throw new ExternalException("SNS publish failed: " + ex.getMessage(), ex);
+      throw new PushFailException("SNS publish failed: " + ex.getMessage(), ex);
```
ExternalException 가 더 최신 코드인데 머지하다가 돌아간 거 같아요!
ExternalException 로 바꿔주시면 감사하겠습니당

JungYoonShin replied:
제가 충돌해결하구,, 몇몇 파일을 같이 커밋을 안해버린 것 같아요... 죄송합니다 ㅠㅠ
요거까지 다시 재커밋 하겠습니당 !!

seoyeonjin commented on src/main/java/com/sopt/push/service/InvalidEndpointCleaner.java:
```diff
@@ -22,7 +22,7 @@ public InvalidEndpointCleaner(
   public void clean(UserTokenInfoDto token) {
 
     userService.deleteUser(token.userId(), token.deviceToken());
-    tokenService.deleteToken(token.userId(), token.deviceToken());
```
이것도 머지하다가 바뀐 거 같아요
`tokenService.deleteToken(token.userId(), token.deviceToken());` 가 맞는 코드입니다!

seoyeonjin commented on src/main/java/com/sopt/push/config/SnsFactory.java:
```diff
@@ -0,0 +1,64 @@
+  private String getPlatformApplicationArn(Platform platform) {
+    String envVar =
+        platform == Platform.IOS ? PLATFORM_APPLICATION_IOS_ENV : PLATFORM_APPLICATION_ANDROID_ENV;
+    String value = System.getenv(envVar);
```
EnvConfig에서 공통으로 getEnv 처리하고 있어서 EnvConfig 이용하도록하면 좋을 것 같아요!

JungYoonShin replied:
넵넵 !! 감사해요

seoyeonjin reviewed (changes requested):
수고하셨습니다!
이전 작업 내용이랑 충돌나서 덮어써진 부분이 조금 있는 것 같아요 !! 그 부분  신경써서 확인해주시면 감사하겠습니당

sung-silver commented on src/main/java/com/sopt/push/config/SnsFactory.java:
```diff
@@ -0,0 +1,64 @@
+            .platformApplicationArn(platformApplicationArn)
+            .token(deviceToken);
+
+    if (userId != null && !userId.isBlank()) {
```
코드 컨벤션인 것 같은데 조건문에 들어가는 boolean expression boolean 변수로 분리해주실 수 있나요??

`boolean isValidUserId = userId != null && !userId.isBlank()`

JungYoonShin replied:
넹 좋습니다 !!

sung-silver commented on src/main/java/com/sopt/push/config/SnsFactory.java:
```diff
@@ -0,0 +1,64 @@
+    String envVar =
+        platform == Platform.IOS ? PLATFORM_APPLICATION_IOS_ENV : PLATFORM_APPLICATION_ANDROID_ENV;
+    String value = System.getenv(envVar);
+    if (value == null || value.isBlank()) {
```
여기도 isInvalidPlatform 이런식으로 표현하면 좋을 것 같아요! 그리고 value라는 변수명이 단순이 값이라고만 해석되어서 명확하게 platform 또는 platformName 이런식으로 바꾸는것도 좋을 것 같습니다!

JungYoonShin replied:
네넵 반영할게용 !!

sung-silver commented on src/main/java/com/sopt/push/dto/LogCreateRequestDto.java:
```diff
@@ -0,0 +1,39 @@
+    Set<String> userIds,
+    Set<String> messageIds,
+    String id) {
+  private static final String NULL = "NULL";
```
해당 변수 Constant 클래스에 빼는 것도 좋아보여요!

JungYoonShin replied:
요거 사용되지 않는 dto인데 제가 안 지웠더라구요,, 리뷰 감사해요 !!

sung-silver commented on src/main/java/com/sopt/push/dto/RegisterHeaderDto.java:
```diff
@@ -0,0 +1,15 @@
+    String transactionId, Services service, Platform platform, Actions action) {
+
+  public RegisterHeaderDto {
+    if ((action == Actions.REGISTER || action == Actions.CANCEL) && platform == null) {
```
여기도 boolean expression 분리해서 구현하면 좋을 것 같습니다!

sung-silver commented on src/main/java/com/sopt/push/lambda/ApiGatewayHandler.java:
```diff
@@ -0,0 +1,335 @@
+      throw new BusinessException(ErrorMessage.INVALID_REQUEST, "Headers missing or invalid.");
+    }
+
+    Map<String, Object> body = parseBody(event);
```
parseBody보다는 parseRequestBody라고 명확하게 변수명 수정하는건 어떨까요??

JungYoonShin replied:
넵 좋아요 !!

sung-silver commented on src/main/java/com/sopt/push/lambda/ApiGatewayHandler.java:
```diff
@@ -0,0 +1,335 @@
+      String actionStr = headers.get("action");
+      String platformStr = headers.get("platform");
+      String transactionId = headers.get("transactionId");
+      String serviceStr = headers.get("service");
```
여기 들어가는 문자열들도 Constant에 상수로 분리하면 좋을 것 같아요!

JungYoonShin replied:
넵 좋아요 반영할게요 ~!

sung-silver commented on src/main/java/com/sopt/push/lambda/ApiGatewayHandler.java:
```diff
@@ -0,0 +1,335 @@
+        if (platformStr == null || platformStr.isBlank()) {
+          throw new BusinessException(
+              ErrorMessage.INVALID_REQUEST, "Platform is required for REGISTER and CANCEL actions");
+        }
```
클린코드 측면에서 중첩 if문을 쓰게되면 읽는 사람이 두번 생각하게 되어서 이부분을 private 메서드로 빼서 checkPlatform() 이런식으로 분리하는 것도 좋을 것 같아요!

JungYoonShin replied:
옹 감사해요 !!

sung-silver commented on src/main/java/com/sopt/push/lambda/ApiGatewayHandler.java:
```diff
@@ -0,0 +1,335 @@
+    Platform platform = request.header().platform();
+    String deviceToken = body.deviceToken();
+    Set<String> userIds = body.userIds();
+    Set<String> logUserIds = Set.of("NULL");
```
앞(LogCreateDto)에서 NULL을 Constant로 뺀다면 여기서 재사용 가능할 것 같아요!

JungYoonShin replied:
감사해용 !! 근데 리팩하면서 LogCreateDto랑 `logUserIds`를 안쓰게 되었어유.. !! 그치만 리뷰 너무너무 고마워용 !!

sung-silver commented on src/main/java/com/sopt/push/lambda/ApiGatewayHandler.java:
```diff
@@ -0,0 +1,335 @@
+      String actualUserId = extractUserIdFromSk(tokenEntity.getSk());
+      String actualDeviceToken = extractDeviceTokenFromPk(tokenEntity.getPk());
+      Platform tokenPlatform = Platform.fromValue(tokenEntity.getPlatform());
+
```
변수선언부니까 붙여쓰면 좋을 것 같아요!

sung-silver commented on src/main/java/com/sopt/push/lambda/ApiGatewayHandler.java:
```diff
@@ -0,0 +1,335 @@
+
+  private void handleSend(ApiGatewayRequestDto request) {
+    SendPushDto body = mapper.convertValue(request.body(), SendPushDto.class);
+
```
여기도 줄바꿈 없으면 좋을 것 같습니다!

sung-silver commented on src/main/java/com/sopt/push/service/DeviceTokenService.java:
```diff
@@ -2,22 +2,35 @@
     this.deviceTokenRepository = deviceTokenRepository;
+    this.userRepository = userRepository;
+    this.snsFactory = snsFactory;
   }
```
여기 롬복으로 대체 가능할 것 같습니다!

JungYoonShin replied:
감사해용 !!

sung-silver reviewed (changes requested):
boolean expression 분리 및 상수 분리 반영 부탁드릴게요!!

JungYoonShin committed:
[REFACTOR] header 정보 상수화

JungYoonShin committed:
[REFACTOR] 필요없는 파일 정리

JungYoonShin committed:
[FEAT] TokenRegisterFacade 구현

- 기존 ApiGatewayHandler에서 토큰 등록하는 부분을 파사드 클래스로 분리

JungYoonShin committed:
[FEAT] EnvConfig env 추가

- 토큰 등록시 aws 엔드포인트 등록을 위한 env를 추가합니다.

JungYoonShin committed:
[REFACTOR] 불필요한 코드 제거

JungYoonShin committed:
[REFACTOR] ApiGatewayHandler 리팩토링

- 토큰 관련 로직은 DeviceTokenService, 유저 관련 로직은 UserService로 이관
- 로깅 방식 @Slf4j으로 수정
- 토큰 등록&삭제, 알림 전송&수정 시 사용하는 DTO를 기존에 존재하는 DTO 재사용하도록 수정
- createRegisterLog, createCancelLogf를 하나의 메서드로 통합

JungYoonShin committed:
[FIX] 잘못된 role 수정

JungYoonShin committed:
[REFACTOR] spotleess 적용

JungYoonShin committed:
[REFACTOR] 필요없는 의존성 및 메서드 제거

JungYoonShin committed:
[FIX] 토큰 등록 실패 시, 토큰 삭제 실패시 성공 응답이 반환되는 문제 수정

JungYoonShin committed:
[REFACTOR] actualUserId 사용하도록 수정

JungYoonShin committed:
[FIX] AppFactory - DeviceTokenService 의존성 정리

sung-silver commented on src/main/java/com/sopt/push/dto/ApiGatewayRequestDto.java:
```diff
@@ -0,0 +1,5 @@
+
+import java.util.Map;
+
+public record ApiGatewayRequestDto(RegisterHeaderDto header, Map<String, Object> body) {}
```
요기 Body는 Map으로 꼭 써야하나요? 요부분 제가 잘 몰라서..ㅎㅎ

JungYoonShin replied:
extractRequest에서 아래 액션을 확인하기 전에 body를 파싱하는데, 어떤 타입으로 변환할지 알 수가 없어서 Map<String, Object>로 받아서, 각 핸들러(handleRegister, handleCancel 등)에서 필요한 타입으로 mapper.convertValue()를 사용해서 변환하고 있습니다 !!

```java
      switch (action) {
        case REGISTER -> handleRegister(request);
        case CANCEL -> handleCancel(request);
        case SEND -> handleSend(request);
        case SEND_ALL -> handleSendAll(request);
        default -> throw new BusinessException(ErrorMessage.INVALID_REQUEST);
      }
```

sung-silver commented on src/main/java/com/sopt/push/lambda/ApiGatewayHandler.java:
```diff
@@ -0,0 +1,310 @@
+import com.sopt.push.config.AppFactory;
+import com.sopt.push.config.ObjectMapperConfig;
+import com.sopt.push.domain.DeviceTokenEntity;
+import com.sopt.push.dto.*;
```
컨벤션 맞추면 좋을 것 같긴하네용!!

sung-silver commented on src/main/java/com/sopt/push/lambda/ApiGatewayHandler.java:
```diff
@@ -0,0 +1,311 @@
+      Actions action = Actions.fromValue(actionStr);
+      Platform platform = fromValue(platformStr);
+
+      if (action == Actions.REGISTER || action == Actions.CANCEL) {
```
여기도 boolean 변수로 빼는거 어떨까요?

```
boolean requiresPlatformCheck =
        action == Actions.REGISTER || action == Actions.CANCEL;

if (requiresPlatformCheck) {
    checkPlatform(platformStr);
}
```

JungYoonShin replied:
넵 알겠습니다 ~!

sung-silver commented on src/main/java/com/sopt/push/lambda/ApiGatewayHandler.java:
```diff
@@ -0,0 +1,311 @@
+import static com.sopt.push.common.StatusCode.BAD_REQUEST;
+import static com.sopt.push.common.StatusCode.INTERNAL_SERVER_ERROR;
+import static com.sopt.push.enums.Platform.fromValue;
+import static com.sopt.push.util.ValidationUtil.validate;
```
validate함수가 뭘 검증하는지 구체적으로 드러나게 함수명을 변경하면 좋을 것 같아요!

seoyeonjin replied:
validate함수를 제가 만들어서,, 잠깐 설명을 드리자면 validate는 spring 에서 `@Valid` 에 해당하는 역할을 하는 함수입니다
따라서 dto 를 말그대로 검증하는 거라서 validate 또는 validateDto 정도가 적당한 이름일 것 같습니다!
(다만 제가 dto 에 `@NotNull `같은 어노테이션을 달아두지 않아서 현재는 큰 의미가 없는 코드입니다.. 후에 리팩하면서 추가하도록 하겠습니다 !!)

JungYoonShin replied:
그럼 validateDto로 수정할까유?? 이게 조금 더 직관적일 것 같긴한데?! 다들 어떤가용 @seoyeonjin @sung-silver

seoyeonjin replied:
좋습니다!

sung-silver replied:
좋습니다!!

sung-silver commented on src/main/java/com/sopt/push/lambda/ApiGatewayHandler.java:
```diff
@@ -0,0 +1,311 @@
+    sendPushFacade.sendPushAll(finalDto);
+  }
+
+  private void checkPlatform(String platformStr) {
```
플랫폼을 체크하는게 ApiGatewayHandler의 책임인지 다시 한 번 생각해보면 좋을 것 같습니다!! 저는 뭔가 Platform(enum)으로 옮기는게 낫지 않나?라는 생각이 들었어요 ㅎㅎ

JungYoonShin replied:
감사해유!!! 근데 코드 보다 보니까 checkPlatform이 필요가 없드라구요.. 이미 Platform enum에 있었어요.. 리뷰 감사해용 !! 덕분에 알게됐습니당

sung-silver commented on src/main/java/com/sopt/push/lambda/ApiGatewayHandler.java:
```diff
@@ -0,0 +1,311 @@
+  }
+
+  private Map<String, Object> parseRequestBody(APIGatewayProxyRequestEvent event) {
+    if (event.getBody() == null) {
```
여기도 boolean 변수로 isInvalidRequest라고 분리하면 좋을 것 같아요!

JungYoonShin replied:
요거 관련해서 궁금한 점이 있어서 낼 구두로 물어보겠습니당 !!

JungYoonShin replied:
요거 나중에 리팩하는걸로 구두로 회의완 !!

sung-silver commented on src/main/java/com/sopt/push/service/TokenRegisterFacade.java:
```diff
@@ -0,0 +1,87 @@
+
+    // 4. UserEntity 저장
+    userService.registerUser(
+        actualUserId, deviceToken, platform.getValue(), endpointArn, subscriptionArn);
```
주석으로 남기기보다 각 과정을 private 메서드로 분리하여 가독성을 높이는건 어떨까요?? 한줄인 부분은 굳이 분리하지 않아도 되겠지만 4줄 이상이면 분리하는 것도 좋을 것 같아요!

JungYoonShin replied:
옹 좋은 것 같아요 !! 반영해볼게용

sung-silver commented on src/main/java/com/sopt/push/service/TokenRegisterFacade.java:
```diff
@@ -0,0 +1,87 @@
+    if (inputUserId == null && tokenUserId.equals("unknown")) {
+      return false;
+    }
+    if (inputUserId != null && inputUserId.equals(tokenUserId)) {
```
"unknown" 상수로 분리하고, 각 조건문 boolean 변수로 빼면 좋을 것 같습니다!

JungYoonShin replied:
요거 코드 보다보니까 조금 이상한 부분이 있어서 전반적으로 수정하게 되었어요 !! 
(사유: unknown의 필요성 x, null인 경우는 dto 단에서 validate 해서 nullable을 막는 것이 좋을 것 같다는 생각이 들었어용)

리뷰 감사합니당 ~!

sung-silver commented on src/main/java/com/sopt/push/service/TokenRegisterFacade.java:
```diff
@@ -0,0 +1,87 @@
+  }
+
+  private boolean shouldSkipRegistration(String existingUserId, String newUserId) {
+    return !changedUserPayload(existingUserId, newUserId);
```
부정연산자를 쓰면 return 값을 직관적으로 이해하기 어려운 것 같아서 boolean 변수로 return 값이 무엇인지 명시적으로 드러내는 것도 좋을 것 같아요!

JungYoonShin replied:
감사해요 !!

sung-silver reviewed (changes requested):
정윤아 한번만 더 확인부탁해용.. ..!!! 고생많았습니다

seoyeonjin commented on src/main/java/com/sopt/push/lambda/ApiGatewayHandler.java:
```diff
@@ -0,0 +1,311 @@
+  private final SendPushFacade sendPushFacade;
+  private final InvalidEndpointCleaner invalidEndpointCleaner;
+  private final HistoryService historyService;
+  private final TokenRegisterFacade tokenRegisterFacade;
```
이건 단순히 컨벤션적인 거긴 한데
InvalidEndpointCleaner 와 TokenRegisterFacade가 비슷한 역할이어서 이름이 비슷하면 더 좋을 거 같아요
```
InvalidEndpointCleaner -> endpointCleanupFacade
TokenRegisterFacade -> endpointRegisterFacade
```
이런식으로 통일하는 건 어떤가요?

JungYoonShin replied:
넵 좋습니다 ! 반영할게용

seoyeonjin commented on src/main/java/com/sopt/push/lambda/ApiGatewayHandler.java:
```diff
@@ -0,0 +1,311 @@
+    };
+  }
+
+  private APIGatewayProxyResponseEvent convertToApiGatewayResponse(
```
응답 만드는 부분을 ResponseUtil로 옮기는 거 어떤가용?? ResponseUtil에서 APIGatewayProxyResponseEvent 생성해서 반환하도록 하면 통일성 측면에서 좋을 것 같아서요!

JungYoonShin replied:
넵 좋아요 반영할게요~!

seoyeonjin commented on src/main/java/com/sopt/push/service/NotificationService.java:
```diff
@@ -14,24 +14,30 @@
     this.snsClient = snsClient;
     this.validator = ValidatorConfig.getValidator();
     this.allTopicArn = envConfig.getAllTopicArn();
+    this.envConfig = envConfig;
```
지금 방식도 괜찮긴 한데, 필요한 값만 내부 필드로 설정해서 사용하는 건 어떤가용?? 
`this.allTopicArn = envConfig.getAllTopicArn();` 처럼 필요한 환경 변수만 가져와서 사용하면 설정값을 확인하기에 좋을 것 같아서 제안드립니당

JungYoonShin replied:
이거 제가 제대로 이해를 못해서 낼 구두로 여쭤보겠습니당 !!

JungYoonShin replied:
넵 반영할게용!!

seoyeonjin commented on src/main/java/com/sopt/push/service/NotificationService.java:
```diff
@@ -114,4 +120,38 @@ public void deleteEndpoint(String endpointArn) {
+  public SubscribeResponse subscribe(String endpointArn) {
+    SubscribeRequest request =
+        SubscribeRequest.builder()
+            .protocol("application")
```
이것도 하드코딩된 값이라 상수화하면 좋을 것 같아요!

JungYoonShin replied:
넵 좋습니다 !

seoyeonjin commented on src/main/java/com/sopt/push/service/TokenRegisterFacade.java:
```diff
@@ -0,0 +1,87 @@
+  private final DeviceTokenService deviceTokenService;
+  private final UserService userService;
+  private final NotificationService notificationService;
+  private final DeviceTokenRepository deviceTokenRepository;
```
혹시 바로 repository를 가져온 이유가 있을까용..? 
다른 의존성들과 일관성을 맞춰서 service 를 통해 접근하는 방식이 좋을 것 같아요

JungYoonShin replied:
넵 알겠습니당 !!

seoyeonjin commented on src/main/java/com/sopt/push/service/TokenRegisterFacade.java:
```diff
@@ -0,0 +1,87 @@
+  private final DeviceTokenRepository deviceTokenRepository;
+
+  public void register(String deviceToken, Platform platform, String userId) {
+    String actualUserId = userId != null ? userId : "unknown";
```
궁금한 점이 하나 있는데용
userId를 "unknown"으로 설정해서 토큰을 등록해야 하는 상황이 있나요??
일반적으로 생각해봤을 때 userId가 매핑되지 않으면 dynamodb 에 토큰을 저장할 이유가 없을 것 같아서요!

JungYoonShin replied:
맞습니당 .. !! 이거 코드가 잘못된 것 같아요.. !! 감사합니다 ~!

seoyeonjin commented on src/main/java/com/sopt/push/service/TokenRegisterFacade.java:
```diff
@@ -0,0 +1,87 @@
+  public void register(String deviceToken, Platform platform, String userId) {
+    String actualUserId = userId != null ? userId : "unknown";
+    String tokenPk = TOKEN_PREFIX + deviceToken;
+    java.util.List<DeviceTokenEntity> existingTokens =
```
사소한 거긴 한데 java.util.List 대신 List 로 사용하면 통일성 측면에서 좋을 것 같습니다!

JungYoonShin replied:
엇 그렇게요.. 수정해둘게요 !!

seoyeonjin commented on src/main/java/com/sopt/push/service/TokenRegisterFacade.java:
```diff
@@ -0,0 +1,87 @@
+      if (shouldSkipRegistration(existingUserId, actualUserId)) {
+        return;
+      }
+      deviceTokenService.deleteToken(existingUserId, deviceToken);
```
제가 정확히 이해한 건진 모르겠는데 userId 변경할 때 dynamodb에서 token 만 지우고 있는 거 맞나용?? 
user정보와 구독 정보는 지우지 않아도 괜찮은지 궁금합니다

seoyeonjin replied:
delete 부분을 외부로 빼는 것도 가능한가요? 제가 코드 로직을 정확히 몰라서 
```
cleaner.clean(); .. 
tokenRegisterFacade.register(deviceToken, platform, userId); 
```
이런식으로 clean을 분리해서 호출하는 게 가능한지 궁금합니다!

JungYoonShin replied:
> 제가 정확히 이해한 건진 모르겠는데 userId 변경할 때 dynamodb에서 token 만 지우고 있는 거 맞나용??
> user정보와 구독 정보는 지우지 않아도 괜찮은지 궁금합니다

맞아요 !! user정보와 구독 정보도 지우는게 맞을 것 같습니당 수정해둘게요 감사합니다 !!
아래 리뷰는 제가 정확히 이해를 못해서 낼 구두로 여쭤볼게용!

seoyeonjin commented on src/main/java/com/sopt/push/service/TokenRegisterFacade.java:
```diff
@@ -0,0 +1,87 @@
+    if (!existingTokens.isEmpty()) {
+      DeviceTokenEntity deviceTokenEntity = existingTokens.get(0);
+      String existingUserId = extractUserId(deviceTokenEntity.getSk());
+      if (shouldSkipRegistration(existingUserId, actualUserId)) {
```
shouldSkipRegistration -> changedUserPayload 로 로직을 2번 감싼 이유가 있을까요? 
shouldSkipRegistration 가 단순히 changedUserPayload의 결과를 반대로 return 하고 있는 것 같아서, 
```
private boolean isSameUser(String existingUserId, String newUserId) {
   ... 
    if ("unknown".equals(existingUserId) && (newUserId == null || "unknown".equals(newUserId))) {
        return true;
    }
   ....
}
```
이런식으로 하나의 메서드로 표현하는 것도 좋을 것 같아요! (성은이는 boolean 변수로 빼는 걸 선호하는 것 같아서 조건문 부분 boolean 변수로 분리하는 것도 좋을 것 같습니당)

JungYoonShin replied:
감사해요 ~~ 수정해두겠습니당

seoyeonjin commented on src/main/java/com/sopt/push/lambda/ApiGatewayHandler.java:
```diff
@@ -0,0 +1,311 @@
+      throw new BusinessException(
+          ErrorMessage.INVALID_REQUEST, "Platform is required for REGISTER and CANCEL actions");
+    }
+  }
```
이것 모에요

seoyeonjin reviewed (commented):
정윤싀 수고하셨습니다 👍
코드 양이 많아서 리뷰도 조금 길어지는 것 같아요
단순히 제안하는 것도 많아서 모두 다 반영하지 않더라도 리소스나 우선순위 고려해서 반영해주시면 좋을 것 같습니다!

Hyobeen-Park reviewed (approved):
멋져요~~!!🚀🚀

JungYoonShin committed:
[REFACTOR] validate > validateDto 메서드명 변경

JungYoonShin committed:
[REFACTOR] InvalidEndpointCleaner & TokenRegisterFacade 통합

- 두 클래스를 하나의 EndpointFacade로 통합

JungYoonShin committed:
[REFACTOR] ApiGatewayHandler - response를 ResponseUtil로 이동

JungYoonShin committed:
[REFACTOR] 토큰 관련 부분 DeviceTokenService로 이동

JungYoonShin committed:
[REFACTOR] SNS_PROTOCOL_APPLICATION 상수화

sung-silver commented on src/main/java/com/sopt/push/service/NotificationService.java:
```diff
@@ -48,7 +57,7 @@ public String platformPush(
           new MessageFactoryDto(
```
이거 messageFactoryDto로 명칭 통일하는게 어떨까요? allTopicPush에서는 dto명을 다 명시하고 있더라구요!!

JungYoonShin replied:
넵 !

JungYoonShin committed:
[REFACTOR] DeviceTokenService - registerToken 메서드 삭제

- 동일한 메서드가 중복되어 registerToken 메서드 삭제

JungYoonShin committed:
[REFACTOR] DeviceTokenRepository에서 사용하지 않는 메서드 삭제

seoyeonjin committed:
[REFACTOR] 코드 포맷팅

seoyeonjin reviewed (approved):
수고하셨습니다 🚀🚀🚀

JungYoonShin committed:
[REFACTOR] 변수명 messageFactoryDto로 수정

JungYoonShin committed:
Merge branch 'feat/#7' of https://github.com/sopt-makers/sopt-push-notification into feat/#7
